private/pkg/storage/storageos/bucket.go

```diff
@@ -404,6 +404,25 @@ func (w *writeObjectCloser) Close() error {
 	return err
 }
 
+// onceError is an object that will only store an error once.
+type onceError struct {
+	err atomic.Value
+}
+
+// Store stores the err.
+func (e *onceError) Store(err error) {
+	e.err.CompareAndSwap(nil, err)
+}
+
+// Load loads the stored error.
+func (e *onceError) Load() error {
+	err, ok := e.err.Load().(error)
+	if !ok {
+		return nil
```

Comment:
This seems like a really bad situation - I need to review my Golang a bit, but if this is non-nil, this means that the thing that was stored was not an error on `e.err`, which would be a system error in itself.

Reply:
yeah, I think for safety purposes you could move `onceError` to a different package so someone within this package can't muck with the `e.err` value directly (instead of using `Store`)?

Reply:
Maybe, but that's not really what I'm saying - I'm saying that you are returning `nil` when `ok == false`, and `ok` should only be false if `e.err.Load()` is not of type `error`, which would be a really bad scenario (somehow something other than an `error` got stored on `e.err`. `ok might also be `false` if `e.err.Load()` is `nil`, but this is where my Golang is failing me - I assume that's the only thing that is meant to be captured here.

I would imagine that if you want to do something like this, you need to do:

```go
errIface := e.err.Load()
if errIface == nil {
  return nil
}
err, ok := errIface.(error) 
if !ok {
  return syserror.Newf("expected error but got %T", errIface)
}
return err
```

Reply:
yeah, that looks more technically correct to me - if you somehow stored something that isn't type `error`, you'd always just get `nil` currently. I'll update to your approach.

Reply:
https://github.com/bufbuild/buf/pull/3443/commits/1a41b945067818b0536a484cef27290843a9fe2d - feel free to change, just preferred the `atomicValue` var name.

Reply:
lgtm - I would probably update bufstream as well